### PR TITLE
[#11694][feat] AutoDeploy: Improve the piecewise CG memory usage

### PR DESCRIPTION
**Pull request by nvchenghaoz**
This PR is to improve the memory usage during the piecewise Cudagraph capturing. 

Several main updates:
1. Use the output buffer as a parameter for the attention operators so that during the piecewise cudagraph, there is no eager attention output allocation. Another advantage for this is the output tensor can be managed by the memory pool, which will hugely improve the memory usage. 
2. Update the piecewise cudagraph control code for the output buffer implementation. And wire the input with the output between dynamic (the ops that cannot be CG captured) and static (can be CG captured) regions. 
3. Some bug fixes and the updates. 

With the change, the memory usage of the piecewise cudagraph is under control and can be used for most cases. 

Also the perf improves a bit due to the changes - 
For Qwen 3.5 isl 1k osl 1k, H100 - 
Screenshot 2026-03-09 at 11 44 13 AM

The next step for piecewise CG is, the final gemm for the prefill is really big and AutoDeploy should enable the gather_logits transform by default. This region is marked as dynamic due to the unknown shape of the gemm. This will be addressed by a follow up task. My plan to fix this is to pad the final gemm to the max_batch_size. That would need some further investigation. #12049 to track the work. 

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Added optional output buffer support across custom operations for more efficient memory usage.
  * Implemented dynamic submodule classification utilities for improved graph compilation.
  * Added batch capacity-based pruning for piecewise compilation buckets.

* **Bug Fixes**
  * Reduced default free GPU memory fraction to prevent OOM errors.

* **Performance Improvements**
  * Enabled piecewise compilation by default for faster inference.
  * Implemented three-phase warmup/capture/replay pipeline with weak reference memory recycling.

### CODE REVIEW

**Review by taylor-yb-lee — APPROVED** [review]

**Comment by nvchenghaoz** [comment]
/bot run --extra-stage "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1" --disable-fail-fast
